### ISSUE TITLE
A way to get a HashFunction backed by a MessageDigest algorithm

### ISSUE THREAD

**Issue by kevinb9n** [google/guava#5990]
Another way to satisfy #3960

**Comment by ritikBhandari** [comment]
Is MessageDigest the same as MD5 algo?

**Comment by lowasser** [comment]
MessageDigest includes, but is not limited to, MD5.

**Comment by ritikBhandari** [comment]
So is HashFunction currently backed by any other algo?

**Comment by ritikBhandari** [comment]
![mdalgo](https://user-images.githubusercontent.com/69508348/160459941-4888f7c8-0b94-4270-aad4-9c33b2f6ab27.png)

Among these, which MD algo do we have to back `HashFunction` with? and which one are we currently using?

**Comment by lowasser** [comment]
`HashFunction` is an interface with many implementations, including those three (`Hashing.md5()`, `Hashing.sha1()`, `Hashing.sha256()`).  Those three are backed by MessageDigest; the others are not.

**Comment by lowasser** [comment]
The point is that we don't expose a utility in _general_ to allow any `MessageDigest` algorithm to be used as a `HashFunction`.

**Comment by ritikBhandari** [comment]
> HashFunction is an interface with many implementations, including those three (Hashing.md5(), Hashing.sha1(), Hashing.sha256()). Those three are backed by MessageDigest; the others are not.

So now do we have to back the other ones as well according to the issue raised?

**Comment by lowasser** [comment]
No, there's no plan to change other implementations of `HashFunction`.  This issue is just about whether or not we should expose a way to make a `HashFunction` from any `MessageDigest`.

**Comment by ritikBhandari** [comment]
Okay! Got it.
Thank you :)

**Comment by nkkolluru** [comment]
@cpovirk Is there any documentation as to what label P3 means?

**Comment by nkkolluru** [comment]
hello! Is the decision to add new API method(s) to `Hashing`, to support creating a `HashFunction` from any `MessageDigest` algorithm taken? Would there be a design process?

**Comment by smartdengjie** [comment]
Hello, I have received your email. Don't worry.

**Comment by vorburger** [comment]
Just FTR for anyone else looking for this in the future, perhaps this analysis could be useful background: 

I've poked around this space while wondering whether to add support for more hash functions [here](https://github.com/enola-dev/enola/blob/fb8abcd985a24a523b362641a33b885513c1e905/java/dev/enola/common/io/hashbrown/Multihashes.java#L29) e.g. for [this](https://docs.enola.dev/use/fetch/#integrity) (related to the `multihash` tagged lines of [this table](https://github.com/multiformats/multicodec/blob/master/table.csv), based on https://github.com/jbenet/random-ideas/issues/1) e.g. for the `MessageDigest` that Guava does not "expose" (intentionally), but which are available at least in the Java 21+ JDK:

* SHA-224, SHA-384, SHA-512/224, SHA-512/256
* SHA3-224, SHA3-256, SHA3-384, SHA3-512 discussed in #3960

I ultimately concluded that, at least for my current application, there was really no need for wanting or needing to do this... but YMMV, of course. It probably boils down to whether you (externally) "have" to support certain hash functions, or whether your application can "choose" which one it wants to support. 

In the former case, it could be "handy" to implement was this FR suggests. But in the latter (and my) case, I've opted to not support SHA-224, SHA-384, SHA-512/224, SHA-512/256 and stick to SHA-256 & SHA-512, only. (BTW all those are sometimes called [SHA-2](https://en.wikipedia.org/wiki/SHA-2) sha2-224, sha2-384, sha2-512-224, sha2-512-256 elsewhere; it took at least me a while to dig this!)

As for SHA3-* (and #3960), after reading "sha3 not being a very good choice" [here](https://github.com/google/guava/issues/3960#issuecomment-1080775346), and noticing https://developers.google.com/tink/protect-data-from-tampering#java (currently) "recommend[ing] the HMAC_SHA256" (=sha2-256), I didn't see any interest in supporting that either by writing a custom Guava `HashFunction` for it.

Just recording this here in case it could be of interest to other Guava users who are investigating how to add more JDK `MessageDigest`-based `HashFunction` implementations in the future - you may or may not actually need & want to do that.
